### PR TITLE
Add p-projectors for Variables and Tensors

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

Add amr::projectors for Variables and Tensors.
This PR only enables support for p-refinement and will ERROR for h-refinement

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/ParallelAlgorithms/Amr/Projectors/Tensors.hpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+#include "NumericalAlgorithms/Spectral/Projection.hpp"
+#include "ParallelAlgorithms/Amr/Protocols/Projector.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace amr::projectors {
+
+/// \brief Update the Tensors corresponding to TensorTags after an AMR
+/// change
```
Comment (nilsvu) [inline src/ParallelAlgorithms/Amr/Projectors/Tensors.hpp:25]: Could you add a bit more detail? At least mention that the tensors are projected between meshes. Also mention `amr::projectors::ProjectVariables`. Also document the template params (TensorTags can be a typelist or pack, etc).

File: src/ParallelAlgorithms/Amr/Projectors/Tensors.hpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+template <size_t Dim, typename... TensorTags>
+struct ProjectTensors : tt::ConformsTo<amr::protocols::Projector> {
+  using return_tags = tmpl::list<TensorTags...>;
```
Comment (nilsvu) [inline src/ParallelAlgorithms/Amr/Projectors/Tensors.hpp:37]: Question: Does the check that all tags are projected work with Variables and subitems?

File: src/ParallelAlgorithms/Amr/Projectors/Variables.hpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+#include "NumericalAlgorithms/Spectral/Projection.hpp"
+#include "ParallelAlgorithms/Amr/Protocols/Projector.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace amr::projectors {
+
+/// \brief Update the Variables corresponding to VariablesTags after an AMR
+/// change
```
Comment (nilsvu) [inline src/ParallelAlgorithms/Amr/Projectors/Variables.hpp:25]: Same request for more docs

File: tests/Unit/ParallelAlgorithms/Amr/Projectors/Test_Variables.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                           Spectral::Quadrature::GaussLobatto};
+  const auto x_old = logical_coordinates(old_mesh);
+  const auto x_new = logical_coordinates(new_mesh);
+  auto var_0 = make_vars(x_old, 1.0);
+  auto var_1 = make_vars(x_old, 2.0);
+  const auto expected_var_0 = make_vars(x_new, 1.0);
+  const auto expected_var_1 = make_vars(x_new, 2.0);
+  amr::projectors::ProjectVariables<
+      Dim, tmpl::list<VariablesTag<0>, VariablesTag<1>>>::
+      apply(make_not_null(&var_0), make_not_null(&var_1), new_mesh,
```
Comment (nilsvu) [inline tests/Unit/ParallelAlgorithms/Amr/Projectors/Test_Variables.cpp]: There's no check here that the projector works with a DataBox and the AMR infrastructure (checks that tags and apply function are compatible, AMR code and apply function are compatible). I think both would be useful.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/ParallelAlgorithms/Amr/Projectors/Tensors.hpp (hunk 2 above)
Comment (kidder) [reply src/ParallelAlgorithms/Amr/Projectors/Tensors.hpp:37]: The DataBox is responsible for handling the subitems.   AMR projectors cannot be called on subitems.

File: tests/Unit/ParallelAlgorithms/Amr/Projectors/Test_Variables.cpp (hunk 4 above)
Comment (kidder) [reply tests/Unit/ParallelAlgorithms/Amr/Projectors/Test_Variables.cpp]: Isn't that the reason for the protocol?  That I have promised that this mutator is usable as an amr::projector?

Comment (nilsvu) [reply tests/Unit/ParallelAlgorithms/Amr/Projectors/Test_Variables.cpp]: Yes, but is that promise tested? :)

**Review by nilsvu — APPROVED** [review]

**Review by knelli2 — APPROVED** [review]

**Comment by knelli2** [comment]
Ignoring gcc-11 Release

**Pull request merged**